platforms/software/ivy/src/main/java/org/gradle/api/publish/ivy/internal/artifact/AbstractIvyArtifact.java

```diff
@@ -16,95 +16,86 @@
 
 package org.gradle.api.publish.ivy.internal.artifact;
 
-import com.google.common.base.Strings;
 import org.gradle.api.internal.tasks.DefaultTaskDependency;
 import org.gradle.api.internal.tasks.TaskDependencyFactory;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.provider.Property;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.provider.ProviderFactory;
+import org.gradle.api.tasks.Optional;
 import org.gradle.api.tasks.TaskDependency;
 
-import javax.annotation.Nullable;
-
 public abstract class AbstractIvyArtifact implements IvyArtifactInternal {
     private final TaskDependency allBuildDependencies;
     private final DefaultTaskDependency additionalBuildDependencies;
-
-    private String name;
-    private String type;
-    private String extension;
-    private String classifier;
-    private String conf;
-
-    protected AbstractIvyArtifact(TaskDependencyFactory taskDependencyFactory) {
+    private final Property<String> nameProperty;
+    private final Property<String> typeProperty;
+    private final Property<String> extensionProperty;
+    private final Property<String> classifierProperty;
+    private final Property<String> confProperty;
+
+    protected AbstractIvyArtifact(
+        TaskDependencyFactory taskDependencyFactory,
+        ProviderFactory providerFactory,
+        ObjectFactory objectFactory
+    ) {
         this.additionalBuildDependencies = new DefaultTaskDependency();
         this.allBuildDependencies = taskDependencyFactory.visitingDependencies(context -> {
             context.add(getDefaultBuildDependencies());
             additionalBuildDependencies.visitDependencies(context);
         });
-    }
-
-    @Override
-    public String getName() {
-        return name != null ? name : getDefaultName();
-    }
-
-    protected abstract String getDefaultName();
 
-    @Override
-    public void setName(String name) {
-        this.name = Strings.nullToEmpty(name);
+        this.nameProperty = objectFactory.property(String.class);
+        this.typeProperty = objectFactory.property(String.class);
+        this.extensionProperty = objectFactory.property(String.class);
+        this.classifierProperty = objectFactory.property(String.class);
+        this.confProperty = objectFactory.property(String.class);
+
+        // those should be lazy because the fields are not yet initialized in child classes
+        getName().set(providerFactory.provider(() -> getDefaultName().getOrNull()));
+        getType().set(providerFactory.provider(() -> getDefaultType().getOrNull()));
+        getExtension().set(providerFactory.provider(() -> getDefaultExtension().getOrNull()));
+        getClassifier().set(providerFactory.provider(() -> getDefaultClassifier().getOrNull()));
+        getConf().set(providerFactory.provider(() -> getDefaultConf().getOrNull()));
```

Comment:
Why even make these `getDefaultX` methods return providers at all if we're not going to treat them as providers? This just makes the PR have more code changes without any real benefit IMO

Reply:
I’m fine the way it is, but to me it does seem we added a bunch of fake laziness and used this code snippet to pretend it’s actually lazy.

I feel like declaring these as providers but not wiring them properly creates a false sense of “provider migration”

do we see a path to wiring the providers properly? Is there a future where the providers will be initialized in the children? If not I don’t see why we should pretend these are providers when we don’t use them properly

Reply:
> different from having an eager property 

How so? Am I missing somewhere where we use these lazily? 

`getName().set(providerFactory.provider(() -> getDefaultName().getOrNull()));`

Is lazy because of the `providerFactory.provider(() -> ...)` 

All I'm saying is there are a bunch of changes in this PR that are not necessary and don't really do much. Having a provider and only calling `.getOrNull` on it means you're not using it as a provider. 

If you see a path forward in the future to removing this extra wrapping/unwrapping of providers that we're doing here, I can see why it might make sense to introduce these changes, but if not I think we're introducing a false sense of provider-ness here

Reply:
In the interest of not dragging this PR out, we can leave it as-is. 

It just seems unnecessary to me